tests/test_run_steps_filter.py

```diff
@@ -19,75 +19,66 @@ def step(name, cmd=False):
-         []],))
-def test_steps_filter(docker_main_and_nonci: UniversumRunner, filters, expected_logs, unexpected_logs):
-    console_out_log = docker_main_and_nonci.run(config, additional_parameters=f"-o console -f='{filters}'")
-    for log_str in expected_logs:
-        assert log_str in console_out_log
-
-    for log_str in unexpected_logs:
-        assert log_str not in console_out_log
-
 
-def test_steps_filter_few_flags(docker_main_and_nonci: UniversumRunner):
-    console_out_log = docker_main_and_nonci.run(config,
-                                                additional_parameters="-o console -f='parent 1:parent 2' -f='!step 1'")
-    for log_str in ["parent 1", "step 2", "parent 2"]:
-        assert log_str in console_out_log
+filters_parametrize_values = (
+    [["parent 1"], ["parent 1", "step 1", "step 2"], ["parent 2"]],
+    [["!parent 2"], ["parent 1", "step 1", "step 2"], ["parent 2"]],
+    [["parent 1:parent 2"], ["parent 1", "step 1", "step 2", "parent 2"], []],
 
-    assert "step 1" not in console_out_log
+    [["parent 1:parent 2:!step 1"], ["parent 1", "step 2", "parent 2"], ["step 1"]],
+    [["step 2"], ["parent 1", "parent 2", "step 2"], ["step 1"]],
+    [["!step 2:parent 1"], ["parent 1", "step 1"], ["parent 2", "step 2"]],
+    [["step :!step 1"], ["parent 1", "parent 2", "step 2"], ["step 1"]],
 
+    [[""], ["parent 1", "parent 2", "parent 1 step 1", "parent 2 step 1", "parent 1 step 2", "parent 2 step 2"], []],
+    [["!"], ["parent 1", "parent 2", "parent 1 step 1", "parent 2 step 1", "parent 1 step 2", "parent 2 step 2"], []],
 
-def test_steps_filter_no_match(tmpdir, capsys):
-    check_filter_no_match(tmpdir, capsys, get_cli_params(tmpdir))
+    [["parent 1:parent 2", "!step 1"], ["parent 1", "step 2", "parent 2"], ["step 1"]]
+)
 
+test_types = ["main", "nonci"]
 
-def test_steps_filter_no_match_nonci(tmpdir, capsys):
-    check_filter_no_match(tmpdir, capsys, nonci_cli_params)
 
+@pytest.mark.parametrize("test_type", test_types)
+@pytest.mark.parametrize("filters, expected_logs, unexpected_logs", filters_parametrize_values)
+def test_steps_filter(tmpdir, capsys, filters, expected_logs, unexpected_logs, test_type):
+    params = get_cli_params(test_type, tmpdir)
+    params.extend(["-o", "console"])
+    for _filter in filters:
+        params.append(f"-f={_filter}")
+    params.extend(["-cfg", get_config_file_path(tmpdir, config)])
 
-def test_config_empty(tmpdir, capsys):
-    check_empty_config_error(tmpdir, capsys, get_cli_params(tmpdir))
-
+    return_code = __main__.main(params)
+    captured = capsys.readouterr()
 
-def test_config_empty_nonci(tmpdir, capsys):
-    check_empty_config_error(tmpdir, capsys, nonci_cli_params)
+    assert return_code == 0
+    for log_str in expected_logs:
+        assert log_str in captured.out
+    for log_str in unexpected_logs:
+        assert log_str not in captured.out
+    assert not captured.err
```

Comment:
I can't find a place where we expect stderr and check its contents. Maybe, using FuzzyCallChecker would suite these tests better?